### PR TITLE
Fix FP S4123 (`no-invalid-await`): Ignore function calls for functions whose definition have JSdoc with a `returns` tag

### LINKED ISSUE #4516: Fix FP S4123 (`no-invalid-await`): Don't flag functions returning promises and message
When calling `getTypeFromTreeNode()` on the node, the returned values has a `.symbol.escapedName` property that is equal to `Promise`

Reproducer:

```js
/**
 * @return {string}
 */
async function foo() {
  await fetch('https://unpkg.com/tailwindcss/dist/tailwind.min.css');
  return 'bar';
}

async function baz() {
  await foo(); // FP
}
```

Improve the message or the rule description: add `make sure that your function is returning a promise`

Proposal: Add `ts.TypeFlags.Void` to the white list

### PR DESCRIPTION
**Pull request by ilia-kebets-sonarsource**
Fixes #4516

### CODE REVIEW

File: packages/jsts/src/rules/S4123/rule.ts [hunk 1]
```diff
@@ -57,6 +61,28 @@ export const rule: Rule.RuleModule = {
...
+  }
+  const callExpression = node.argument;
+  const tsCallExpr = services.esTreeNodeToTSNodeMap.get(callExpression.callee as TSESTree.Node);
+  const tc = services.program.getTypeChecker();
+  const symbol = tc.getSymbolAtLocation(tsCallExpr);
+  const declarations = symbol?.declarations;
+  if (declarations?.some(d => hasJsDoc(d))) {
+    return true;
+  }
+  return false;
```
Comment (yassin-kammoun-sonarsource) [inline packages/jsts/src/rules/S4123/rule.ts:79]:
```suggestion
  if (node.argument.type !== 'CallExpression') {
    return false;
  }
  const signature = getSignatureFromCallee(node.argument, services);
  return signature?.declaration && hasJsDoc(signature.declaration);
```

Comment (yassin-kammoun-sonarsource) [reply packages/jsts/src/rules/S4123/rule.ts:79]: You will need to import `getSignatureFromCallee`.

File: packages/jsts/src/rules/S4123/rule.ts [hunk 2]
```diff
@@ -57,6 +61,28 @@ export const rule: Rule.RuleModule = {
...
+
+  function hasJsDoc(declaration: ts.Declaration & { jsDoc?: ts.JSDoc[] }) {
+    return declaration.jsDoc && declaration.jsDoc.length > 0;
```
Comment (yassin-kammoun-sonarsource) [inline packages/jsts/src/rules/S4123/rule.ts:82]: Should we be more accurate and check that there is a `@return` or `@returns` tag?

File: packages/jsts/src/rules/S4123/rule.ts [hunk 3]
```diff
@@ -34,7 +35,10 @@ export const rule: Rule.RuleModule = {
     const services = context.sourceCode.parserServices;
     if (isRequiredParserServices(services)) {
       return {
-        AwaitExpression: (node: estree.Node) => {
+        AwaitExpression: (node: estree.AwaitExpression) => {
+          if (isException(node, services)) {
+            return;
+          }
```
Comment (yassin-kammoun-sonarsource) [inline packages/jsts/src/rules/S4123/rule.ts:41]: Could you also include in the if-statement below another condition `!isPromise(awaitedType)` where `isPromise` checks what we discussed together last week about the Promise type and mentioned in the ticket:

> symbol.escapedName property that is equal to Promise

**Review by yassin-kammoun-sonarsource — CHANGES_REQUESTED** [review]

File: packages/jsts/src/rules/S4123/rule.ts (hunk 3 above)
Comment (ilia-kebets-sonarsource) [reply packages/jsts/src/rules/S4123/rule.ts]: We will not do this.

As discussed, the current check implemented in isThennable seems to be englobing this new proposed verification:

```ts
function isPromise(type: ts.Type) {
  return type.symbol?.escapedName === 'Promise';
}
```

In practice, adding this check did not yield any additional issues on ruling, and we have not been able to define a test case for it.
However, when replacing `isThennable()` by this proposed check, we obtain additional FPs in the ruling tests:

```
# javascript-S4123.json
+"postgraphql:src/postgraphql/__tests__/postgraphqlIntegrationMutations-test.js": [
+48
+],
+"postgraphql:src/postgraphql/__tests__/postgraphqlIntegrationQueries-test.js": [
+55
+],
+"postgraphql:src/postgres/inventory/collection/__tests__/PgCollection-test.js": [
+81,
+170,
+177,
+360,
+361
+],
```

```
# typescript-S4123.json
+{
 │+"postgraphql:src/__tests__/utils/withPgClient.ts": [
 │+24,
 │+25,
 │+29,
 │+33,
 │+50
 │+],
 │+"postgraphql:src/postgraphql/schema/procedures/createPgProcedureMutationGqlFieldEntry.ts": [
 │+117
 │+],
 │+"postgraphql:src/postgraphql/schema/procedures/createPgProcedureObjectTypeGqlFieldEntry.ts": [
 │+61
 │+],
 │+"postgraphql:src/postgraphql/schema/procedures/createPgProcedureQueryGqlFieldEntry.ts": [
 │+58
 │+],
 │+"postgraphql:src/postgraphql/watch/watchPgSchemas.ts": [
 │+40,
 │+48,
 │+53
 │+],
 │+"postgraphql:src/postgraphql/withPostGraphQLContext.ts": [
 │+58,
 │+79,
 │+180
 │+],
 │+"postgraphql:src/postgres/introspection/introspectDatabase.ts": [
 │+26
 │+],
 │+"postgraphql:src/postgres/inventory/collection/PgCollection.ts": [
 │+177
 │+],
 │+"postgraphql:src/postgres/inventory/collection/PgCollectionKey.ts": [
 │+190,
 │+246,
 │+280
 │+],
 │+"postgraphql:src/postgres/inventory/paginator/PgPaginator.ts": [
 │+37
 │+],
 │+"postgraphql:src/postgres/inventory/paginator/PgPaginatorOrderingAttributes.ts": [
 │+102,
 │+130,
 │+147
 │+],
 │+"postgraphql:src/postgres/inventory/paginator/PgPaginatorOrderingOffset.ts": [
 │+138
 │+]
```

Comment (yassin-kammoun-sonarsource) [reply packages/jsts/src/rules/S4123/rule.ts]: 👍

**Review by yassin-kammoun-sonarsource — APPROVED** [review]
LGTM!

**Pull request merged**